jax/_src/numpy/lax_numpy.py

```diff
@@ -2209,19 +2209,38 @@ def _convert_to_array_if_dtype_fails(x: ArrayLike) -> ArrayLike:
   else:
     return x
 
-
 @util.implements(getattr(np, "astype", None), lax_description="""
 This is implemented via :func:`jax.lax.convert_element_type`, which may
 have slightly different behavior than :func:`numpy.astype` in some cases.
 In particular, the details of float-to-int and int-to-float casts are
 implementation dependent.
 """)
-def astype(x: ArrayLike, dtype: DTypeLike | None, /, *, copy: bool = True) -> Array:
-  del copy  # unused in JAX
+def astype(x: ArrayLike, dtype: DTypeLike | None, /, *, copy: bool = True, device: xc.Device | Sharding | None = None) -> Array:
   if dtype is None:
     dtype = dtypes.canonicalize_dtype(float_)
   dtypes.check_user_dtype_supported(dtype, "astype")
-  return lax.convert_element_type(x, dtype)
+  src_dtype = x.dtype if hasattr(x, "dtype") else dtypes.dtype(x)
+  if (
+    src_dtype is not None
+    and dtypes.isdtype(src_dtype, "complex floating")
+    and dtypes.isdtype(dtype, ("integral", "real floating"))
+  ):
+    raise ValueError(
```

Comment:
This is a breaking change: currently e.g. `jnp.astype(1 + 1j, float)` raises a `ComplexWarning` and returns `jnp.array([1.0])`.

Is erroring here required by the Array API spec?

Reply:
> Is erroring here required by the Array API spec?

Yes.

> This is a breaking change: currently e.g. `jnp.astype(1 + 1j, float)` raises a ComplexWarning and returns `jnp.array([1.0])`.

In that case, I can move the error handling to `array_api.astype` and include a `DeprecationWarning` here in addition to the `ComplexWarning`.